crates/egui/src/atomics/atomic_layout.rs

```diff
@@ -0,0 +1,487 @@
+        let Self {
+            sized_atomics,
+            frame,
+            fallback_text_color,
+            response,
+            grow_count,
+            desired_size,
+            align2,
+            gap,
+        } = self;
+
+        let inner_rect = response.rect - self.frame.total_margin();
+
+        ui.painter().add(frame.paint(inner_rect));
+
+        let width_to_fill = inner_rect.width();
+        let extra_space = f32::max(width_to_fill - desired_size.x, 0.0);
+        let grow_width = f32::max(extra_space / grow_count as f32, 0.0).floor_ui();
+
+        let aligned_rect = if grow_count > 0 {
+            align2.align_size_within_rect(Vec2::new(width_to_fill, desired_size.y), inner_rect)
+        } else {
+            align2.align_size_within_rect(desired_size, inner_rect)
+        };
+
+        let mut cursor = aligned_rect.left();
+
+        let mut response = AtomicLayoutResponse::empty(response);
+
+        for sized in sized_atomics {
+            let size = sized.size;
+            let growth = if sized.is_grow() { grow_width } else { 0.0 };
+
+            let frame = aligned_rect
+                .with_min_x(cursor)
+                .with_max_x(cursor + size.x + growth);
+            cursor = frame.right() + gap;
+
+            let align = Align2::CENTER_CENTER;
+            let rect = align.align_size_within_rect(size, frame);
+
+            match sized.kind {
+                SizedAtomicKind::Text(galley) => {
+                    ui.painter().galley(rect.min, galley, fallback_text_color);
+                }
+                SizedAtomicKind::Image(image, _) => {
+                    image.paint_at(ui, rect);
+                }
+                SizedAtomicKind::Custom(id, _) => {
+                    debug_assert!(
+                        !response.custom_rects.iter().any(|(i, _)| *i == id),
+                        "Duplicate custom id"
+                    );
+                    response.custom_rects.push((id, rect));
+                }
+                SizedAtomicKind::Empty => {}
+            }
+        }
+
+        response
+    }
+}
+
+/// Response from a [`AtomicLayout::show`] or [`AllocatedAtomicLayout::paint`].
+///
+/// Use the `custom_rects` together with [`AtomicKind::Custom`] to add child widgets to a widget.
+///
+/// NOTE: Don't `unwrap` rects, they might be empty when the widget is not visible.
+#[derive(Clone, Debug)]
+pub struct AtomicLayoutResponse {
+    pub response: Response,
+    // There should rarely be more than one custom rect.
+    custom_rects: SmallVec<[(Id, Rect); 1]>,
+}
+
+impl AtomicLayoutResponse {
+    pub fn empty(response: Response) -> Self {
+        Self {
+            response,
+            custom_rects: Default::default(),
+        }
+    }
+
+    pub fn custom_rects(&self) -> impl Iterator<Item = (Id, Rect)> + '_ {
+        self.custom_rects.iter().copied()
+    }
+
+    pub fn get_rect(&self, id: Id) -> Option<Rect> {
```

Comment:
https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

```suggestion
    pub fn rect(&self, id: Id) -> Option<Rect> {
```